[GitRepository] Use multi-process file locking for git actions

NickLaMuro opened the pull request:
Adds `GitRepository#git_transaction` method as well as the following private helper methods and constants:

```ruby
GitRepository::LOCKFILE_DIR  # Where the locks are stored on the appliance

repo = GitRepository.new
repo.git_lock_filename       # name of the lockfile for this record
repo.acquire_git_lock        # fetches a lock for the instance (one per)
repo.release_git_lock        # releases existing lock for other processes
```

`GitRepository#git_transaction` is meant to be a multi-purpose method for dealing with any file system changes so that only one process is able to interact with it at a time.  According to the docs (`ri File#flock`), `File#flock(FILE::LOCK_EX)` should block until the lock is available for the process, so the first process to acquire the lock will process first.

The method is also designed so that nested `GitRepository#git_transaction` calls will no-op and not release the lock pre-maturely.  This is important for running multiple actions that by design need to happen by a single process, but the sub actions are also configured to use the lock.

Implementation of this method will be done in a followup commit, but the method is also meant to be a public interface so that when existing methods implement this method, locks will work just fine with user defined `git_transaction` blocks as well as with the `GitRepository` public API.


Links
-----

* Addressing ["`git clone` not processes safe" concern](https://github.com/ManageIQ/manageiq-design/issues/45#issuecomment-514372257) from RFE
* `AndibleRunner` for `EmbeddedAnsible` RFE:  https://github.com/ManageIQ/manageiq-design/issues/45


Steps for Testing/QA
--------------------

¯\\(°\_o)/¯

~~But in all seriousness, this is really hard (impossible) to test since it requires running two processes and hitting a process timing where this would become an issue.  That said, hopefully it doesn't cause the existing specs to break.~~

While not ideal, I have put together a test script for this:

https://gist.github.com/NickLaMuro/20dc0795410abf448402c54f707a59dc

Running that seemed to prove that things were working as expected, in addition to the `Thread` based test I provided in a comment below:

https://github.com/ManageIQ/manageiq/pull/19074#issuecomment-516624369

NickLaMuro commented on app/models/git_repository.rb:
```diff
@@ -128,15 +154,19 @@ def refresh_tags
     @worktree ||= begin
-      clone_repo unless Dir.exist?(directory_name)
-      fetch_worktree
+      git_transaction do
```
Unsure if this is needed... I don't think `GitRepository#worktree` is ever accessed without `GitRepository#with_worktree`, but maybe I am missing something.

If nothing else, it at least tests that the "top level only" transaction thing works (the test suite should exercise this cod) from `EmbeddedAnsible::AutomationManager::ConfigurationScriptSource`).

Fryguy replied:
I think only the clone line + directory check needs the lock around it.  I'd be ok if the Dir.exist? check lived inside clone_repo method so that you can encapsulate the lock in that one method.

NickLaMuro commented:
@miq-bot add_label core, embedded ansible

(not sure if this should be a `bug` or an `enhancement`...)

@Fryguy and @carbonin please review

carbonin commented on app/models/git_repository.rb:
```diff
@@ -155,6 +185,34 @@ def handling_worktree_errors
+  end
+
+  def acquire_git_lock
+    return false if defined? @git_lock
```
Maybe it would be better to maintain a handle to the lock file as an instance variable constantly?
I would suggest making `lock_file` a memoized instance method that opens the file and returns the handle. Then you can rely on the `.flock(File::LOCK_EX)` being a no-op if you already hold the lock rather than doing this check.

NickLaMuro replied:
@carbonin I pretty sure agree with your first comment, but not so much with the second.

I need the usage of this in `git_transaction` to return `false` when this exists so that the `should_unlock` works in the `ensure` (or rather, doesn't trigger the unlock early), and I don't think what you are suggesting helps address that.

Maybe a more concrete example will handle prove that I am missing something in your explanation?  Seems like just adding `&& !@git_lock.nil?` to the end of this would have the same effect as what I think you are trying to suggest.

NickLaMuro commented on app/models/git_repository.rb:
```diff
@@ -155,6 +185,34 @@ def handling_worktree_errors
+  end
+
+  def release_git_lock
+    return unless defined? @git_lock
```
This one is a little different than the above.  I am merely using this `defined?` as a guard against a `nil` error, but also against a warning.  You are right in that it shouldn't **just** be a `defined?` check, but also a `nil` check.  However, if it is just the `nil` check, when you run with warnings turned on, you get something like this:

```console
$ cat example_instance_variable_warning.rb
class Foo
  def self.bar
    puts "bar" if @baz
  end
end

Foo.bar
$ ruby -w example_instance_variable_warning.rb
example_instance_variable_warning.rb:3: warning: instance variable @baz not initialized
```

Which we might currently avoid in prod and specs, but I don't want to add to the mess even if we do.

NickLaMuro commented:
> This makes git repos process safe, but not thread safe. Should we also wrap the `acquire_git_lock` method in a mutex?

@carbonin after some research, I think this process already makes it thread safe in the "process" of making it "process safe":

```ruby
class FileLock
  def self.id
    if STDOUT.tty?
      @thread_colors ||= {}
      @current_id    ||= 30

      @thread_colors[Thread.current.object_id] ||= "\e[#{@current_id += 1}m#{Thread.current.object_id}\e[0m"
    else
      Thread.current.object_id
    end
  end

  def self.run
    puts "  > #{id} starting to run..."
    sleep 1

    File.open ".lockfile", "w" do |lockfile|
      lockfile.flock(File::LOCK_EX)
      puts "  > #{id} has obtained the lock!"
      sleep 1
      lockfile.flock(File::LOCK_UN)
      puts "  > #{id} has released the lock!"
    end
  end
end

puts "starting..."

threads = []
3.times do
  threads << Thread.new { FileLock.run }
end

threads.each(&:join)

puts "completed!"
```


```console
$ ruby flock_thread_safety.rb
starting...
  > 70269252667060 starting to run...
  > 70269252667200 starting to run...
  > 70269252667340 starting to run...
  > 70269252667200 has obtained the lock!
  > 70269252667200 has released the lock!
  > 70269252667340 has obtained the lock!
  > 70269252667340 has released the lock!
  > 70269252667060 has obtained the lock!
  > 70269252667060 has released the lock!
completed!
```

NickLaMuro commented:
@Fryguy @carbonin when ever you have time, I think I am pretty much set with this one.

I have provided some tests scripts and run info exercising this, but not sure it is worth trying to add to the test suite (I can if you want).

Also, my concern from my last commit is still unanswered:

> `GitRepository#worktree` probably doesn't need `#git_transaction` since it should only be called from `#with_worktree`, but want to confirm with others.

Honestly, doesn't seem to be hurting anything, but curious if you have strong opinions on it staying.  Once a decision has been decided on, I will update that commit and un-`[WIP]`

Fryguy commented on app/models/git_repository.rb:
```diff
@@ -70,9 +71,14 @@ def self_signed_cert_cb(_valid, _host)
 
+  # NOTE:  Please make sure to use this whenever you implement a method that is
+  # modifying the `.git` data (clone, fetch, checking, etc.) so that we can be
+  # sure this happens safely between processes.
```
Rugged handles locking internally, so 2 separate instanced of a Rugged::Repo doing a fetch will not collide with each other.  The only thing we need to handle are clones because that's occurring outside of a "repo" context.

Fryguy replied:
I think if there are multiple processes looking at the same repo, I'd also like to *not* serialize them, allowing them to execute git stuff in parallel.

NickLaMuro replied:
Well this would have been good to know prior to doing all this work...

NickLaMuro replied:
Okay, "salty Nick" has left the building...

Does it make sense to then reuse the locking code that is in `GitWorktree`/`Rugged` (honestly still not fully up to speed on which does it currently):

https://github.com/ManageIQ/manageiq/blob/master/lib/git_worktree.rb#L366-L391

or stick with what I have done here?

carbonin replied:
I think the linked locking bits would still have to be after the clone, so it doesn't help us in this case.

Fryguy commented on app/models/git_repository.rb:
```diff
@@ -86,6 +92,26 @@ def update_repo
+  # This means that you can surround a couple of actions with this method, and
+  # the lock will only be enforced on the top level.
+  #
+  # Return value is the result of the yielded block
```
This all feels like a lot of comment to say "lock the directory to avoid clone collisions".  I don't think we need this entire comment.  locks themselves are pretty clear to most readers.

NickLaMuro replied:
Hey man, I never said that my 📚 were concise, or even grammatically correct (holy typos batman... I was clearly out of it when I wrote this...).

That said, I think at least some things are worth noting based on how I implement it:

The nested transactions bit is important because if it wasn't there you would lock in the outer `git_transaction` and deadlock in the nested one.  I also wanted devs to know that this was designed ensure that the lock was only released in the outer most `git_transaction`, because releasing early would have defeated the purpose.

That said, based on your [other comment](https://github.com/ManageIQ/manageiq/pull/19074#discussion_r309870129), I am more questioning my life choices and less concerned about the length/validity of my comment here...

(in otherwords, this comment being relevant depends first if this implementation is even necessary, and determining if it is concise enough can be answered only after)

Fryguy commented on app/models/git_repository.rb:
```diff
@@ -155,6 +183,36 @@ def handling_worktree_errors
+    @git_lock.flock(File::LOCK_EX)                         # block waiting for lock
+    @git_lock.write("#{Process.pid} - #{Time.zone.now}\n") # for debugging
+    @git_lock.flush                                        # write current data
+    @git_lock.truncate(@git_lock.pos)                      # clean up remaining chars
```
Why is this last line needed if you've flushed?

NickLaMuro replied:
Because if you don't, you could end up with something like this:

```ruby
File.open "tmp/fake.lock", File::RDWR | File::CREAT do |f|
  f.puts "really_really_long_pid - consistent_length_timestamp"
  f.flush
end

puts
puts "First Write"
puts "-----------"
puts
puts File.read "tmp/fake.lock"

File.open "tmp/fake.lock", File::RDWR | File::CREAT do |f|
  f.puts "short_pid - consistent_length_timestamp"
  f.flush
end

puts
puts "Second write"
puts "------------"
puts
puts File.read "tmp/fake.lock"


# $ ruby tmp/test_ruby_file_truncation.rb
#
# First Write
# -----------
#
# really_really_long_pid - consistent_length_timestamp
#
# Second write
# ------------
#
# short_pid - consistent_length_timestamp
# th_timestamp
```

Where the `"th_timestamp"` is the leftovers from the previous file write operation.

`File#flush` is there just to make sure we push to disk as soon as possible and not let it buffer in memory until we close the file (pretty sure), and `File#truncate` clears out any data after the cursor position (in this case).

Fryguy replied:
Wow that is funky.  I would have expected the flush to ensure that doesn't happen.  Another option is to use `f.sync = true` maybe?

Fryguy replied:
Ohhhhh I see...it took me a while to understand but this opens an existing file for random access, as opposed to the usual open a file for write and auto-truncate.

Fryguy commented on app/models/git_repository.rb:
```diff
@@ -155,6 +183,36 @@ def handling_worktree_errors
+  end
+
+  def release_git_lock
+    return if !defined?(@git_lock) || @git_lock.nil?
```
Do you need the defined check if you're going to check for nil anyway?

carbonin replied:
Maybe it would be easier to make a `#git_lock` attr_accessor?

NickLaMuro replied:
> Maybe it would be easier to make a `#git_lock` attr_accessor?

@carbonin so I did respond to your earlier comments about this already:

https://github.com/ManageIQ/manageiq/pull/19074#discussion_r308969083
https://github.com/ManageIQ/manageiq/pull/19074#discussion_r308967973

However, I think I might have been misguided in my explanation or what you were suggestion.  I will take a look into this a bit later tonight, and I think that will probably help with my original issue.

Fryguy commented on app/models/git_repository.rb:
```diff
@@ -155,6 +183,36 @@ def handling_worktree_errors
+    @git_lock.close
+
+    # TODO:  unlink?  Unsure how to handle this if other processes are waiting
+    # on this though...
```
I don't think we need to unlock because the next process is going to reopen the file.

carbonin replied:
unlock or unlink @Fryguy ?

NickLaMuro replied:
> I don't think we need to unlock because the next process is going to reopen the file.

Granted, not an expert on `File.flock`, but all of the files are "waiting on" or "acquiring" the lock AFTER they have opened the file.  So my thought here was that we would be pulling the rug from under the other processes that are waiting at their own `@git_lock.flock(File::LOCK_EX)` calls for the one with the lock to finish, and if the process with the lock deletes the file when releasing, I am pretty sure that would cause an IO error.

carbonin replied:
Yeah, this was one of the reasons I was suggesting we just leave the file around. I don't think we ever really have to remove is unless the repo is deleted.

Fryguy replied:
> unlock or unlink @Fryguy ?

Typo fail...yes, unlink

Fryguy replied:
Agreed...don't unlink the file for all of the reasons mentioned.

Fryguy commented on app/models/git_repository.rb:
```diff
@@ -155,6 +183,36 @@ def handling_worktree_errors
   end
 
+  def git_lock_filename
+    @git_lock_filename ||= LOCKFILE_DIR.join(id.to_s)
```
Do we need a method + instance variable when there is only one usage of this?

NickLaMuro replied:
No.  Did it for readability mostly.

Fryguy commented on app/models/git_repository.rb:
```diff
@@ -155,6 +183,36 @@ def handling_worktree_errors
+
+    @git_lock = File.open(git_lock_filename, File::RDWR | File::CREAT, 0o644)
+    @git_lock.flock(File::LOCK_EX)                         # block waiting for lock
+    @git_lock.write("#{Process.pid} - #{Time.zone.now}\n") # for debugging
```
If we don't unlink the file later, will this just keep appending since you've opened RDWR?

NickLaMuro replied:
All usages of the this have their cursors at 0, so when they are writing, they are each starting at 0, and not appending on top of each other.

See https://github.com/ManageIQ/manageiq/pull/19074/files#r309950154

Fryguy commented on app/models/git_repository.rb:
```diff
@@ -70,9 +71,14 @@ def self_signed_cert_cb(_valid, _host)
   def with_worktree
-    handling_worktree_errors do
-      yield worktree
+    git_transaction do
```
So this one should be dropped, IMO.

Fryguy replied:
I still don't think this should lock here as this effectively serializes every operation, and we only want to serialize the clones

NickLaMuro commented:
@carbonin @Fryguy Sorry for the delay on this, but I did want to test this properly (manually) before I got back.  Changed 3 things:

- Removed the `.unlink` comment
- Added a comment amount thread safety issue (when sharing an instance)
- Made the `@git_lock` an `attr_reader`

In regards to the `@git_lock` change to an `attr_reader`, it worked as expected when running my [test script](https://gist.github.com/NickLaMuro/20dc0795410abf448402c54f707a59dc) with `RUBYOPT="-w"`, and there even is a good example of this not being done properly with `decendent_loader.rb` in the resulting screenshot:

![git_transaction_example_with_dash_w](https://user-images.githubusercontent.com/314014/62483728-06d69800-b77e-11e9-82dd-fa034f99bcb4.png)

(and I saved you from the almost 800 lines of warnings preceding this output...)

carbonin commented:
Looks good to me. Will let @Fryguy have the last say here.

Fryguy commented:
The two things I was requesting were not done yet...that is, remove the git_transaction from with_worktree and only do git_transaction from within clone_repo (also putting the Dir.exists? call inside of clone_repo.

I want to avoid the locking overhead for every operation when it's only needed for when two processes compete for a clone.

NickLaMuro commented:
> I want to avoid the locking overhead for every operation when it's only needed for when two processes compete for a clone.

I was under the impression after our (out of band) meeting on Friday that we had come to the consensus that it was not a problem leaving it in as is, but apparently not.

Will change in a bit.

NickLaMuro commented:
@Fryguy comments addressed
